t/protocol_directnic.pl

```diff
@@ -0,0 +1,231 @@
+        wantips => {h1 => {wantipv4 => '192.0.2.1'}},
+        wantstatus => {
+            h1 => {'status-ipv4' => 'good', 'ipv4' => '192.0.2.1', 'mtime' => $ddclient::now},
+        },
+        wantlogs => [
+            {label => 'SUCCESS', ctx => ['h1'], msg => qr/IPv4/},
+        ],
+    },
+    {
+        desc => 'IPv4, single host, failed',
+        cfg => {h1 => {url => "$hostname/dns/gateway/bad_token/"}},
+        wantips => {h1 => {wantipv4 => '192.0.2.1'}},
+        wantstatus => {
+            h1 => {'status-ipv4' => 'failed'},
+        },
+        wantlogs => [
+            {label => 'FAILED', ctx => ['h1'], msg => qr/There was an error updating your record/},
+        ],
+    },
+    {
+        desc => 'IPv4, single host, bad',
+        cfg => {h1 => {url => "$hostname/bad/path/"}},
+        wantips => {h1 => {wantipv4 => '192.0.2.1'}},
+        wantstatus => {
+            h1 => {'status-ipv4' => 'bad'},
+        },
+        wantlogs => [
+            {label => 'FAILED', ctx => ['h1'], msg => qr/response is not a JSON object: unexpected response body/},
+        ],
+    },
+    {
+        desc => 'IPv4, single host, unexpected response',
+        cfg => {h1 => {url => "$hostname/unexpected/path/"}},
+        wantips => {h1 => {wantipv4 => '192.0.2.1'}},
+        wantstatus => {
+            h1 => {'status-ipv4' => undef},
+        },
+        wantlogs => [
+            {label => 'FAILED', ctx => ['h1'], msg => qr/400 Bad Request/},
+        ],
+    },
+    {
+        desc => 'IPv4, multiple hosts, multiple good',
+        cfg => {
+            h1 => {urlv4 => "$hostname/dns/gateway/abc/"},
+            h2 => {urlv4 => "$hostname/dns/gateway/def/"},
+        },
+        wantips => {
+            h1 => {wantipv4 => '192.0.2.1'},
+            h2 => {wantipv4 => '192.0.2.2'},
+        },
+        wantstatus => {
+            h1 => {'status-ipv4' => 'good', 'ipv4' => '192.0.2.1', 'mtime' => $ddclient::now},
+            h2 => {'status-ipv4' => 'good', 'ipv4' => '192.0.2.2', 'mtime' => $ddclient::now},
+        },
+        wantlogs => [
+            {label => 'SUCCESS', ctx => ['h1'], msg => qr/IPv4/},
+            {label => 'SUCCESS', ctx => ['h2'], msg => qr/IPv4/},
+        ],
+    },
+    {
+        desc => 'IPv4, multiple hosts, mixed success',
+        cfg => {
+            h1 => {urlv4 => "$hostname/dns/gateway/bad_token/"},
+            h2 => {urlv4 => "$hostname/dns/gateway/def/"},
+        },
+        wantips => {
+            h1 => {wantipv4 => '192.0.2.1'},
+            h2 => {wantipv4 => '192.0.2.2'},
+        },
+        wantstatus => {
+            h1 => {'status-ipv4' => 'failed'},
+            h2 => {'status-ipv4' => 'good', 'ipv4' => '192.0.2.2', 'mtime' => $ddclient::now},
+        },
+        wantlogs => [
+            {label => 'FAILED', ctx => ['h1'], msg => qr/There was an error updating your record/},
+            {label => 'SUCCESS', ctx => ['h2'], msg => qr/IPv4/},
+        ],
+    },
```

Comment:
Because this protocol doesn't update multiple hosts at the same time I don't think these test cases are necessary.